### PR TITLE
Monte carlo Model

### PR DESCRIPTION
**Pull request by IanAndersenNg**
## Description

A monte carlo model with epsilon greedy policy.

Below are some of the definition of the model:

- Episode: A complete run of the model from the start to the end of the training data.
- Epsilon: The probability of choosing a random action instead of the best action for exploring instead of exploiting.
- Portfolio Value: init to be 100 at first and then it will be updated based on the reward on each episode to track the value.

## Testing Results
Example Q table

epsilon greedy behaviour is tested also 

Screenshot 2024-12-20 at 13 51 04

### CODE REVIEW

File: models/Model.py [hunk 1]
```diff
@@ -1,10 +1,12 @@
 
 class Model:
-    def __init__(self, state_space, action_space, gamma=0.99):
+    def __init__(self, state_space, action_space, gamma=0.99, environment = None):
         self.state_space = state_space
         self.action_space = action_space
         self.gamma = gamma
         self.policy = self.initialize_policy()
+        self.environment = environment
+        self.action_set = [[25, 75], [50,50], [75,25], [0,100], [100,0]]
```
Comment (yuhsuanyang) [inline models/Model.py]: `action_set` can be the same as `action_set`

File: models/Model.py [hunk 2]
```diff
@@ -1,10 +1,12 @@
```
Comment (yuhsuanyang) [inline models/Model.py]: Can we rename this file as `models.py` follow by the naming convention of python?

File: models/Model.py (hunk 1 above)
Comment (EdTengHR) [inline models/Model.py]: might be redundant since action space is already defined, but if the code works anyway this can be ignored

File: models/MonteCarlo.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+- Epsilon: The probability of choosing a random action instead of the best action for exploring instead of exploiting.
+- Portfolio Value: init to be 100 at first and then it will be updated based on the reward on each episode to track the value.
+
+'''
+
+class MonteCarloRL(Model):
+    def __init__(self, cleaned_data, actions, n_episodes=1000, epsilon=0.1):
+        super().__init__(state_space=None, action_space=actions)
+        self.cleaned_data = cleaned_data
+        self.actions = actions
```
Comment (EdTengHR) [inline models/MonteCarlo.py]: seems like we can just use action_space from the parent instead of defining a new variable, but again as long as it doesn't impact correctness i guess this can be a secondary concern

File: models/MonteCarlo.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+     
+                nasdaq_return = self.nasdaq_train[t]
+                msci_return = self.msci_train[t]
+
+                reward = self.calculate_reward(action, nasdaq_return, msci_return)
+                portfolio_value *= (1 + reward / 100)
+                episode_rewards.append(reward)
+
+            avg_reward = np.mean(episode_rewards)
+            if episode == 0 or avg_reward > results[best_action_index]['average_reward']:
```
Comment (EdTengHR) [inline models/MonteCarlo.py]: this code looks a little odd to me. results seems to be a list of dicts containing each episode's results, but is accessed by the best action index, which corresponds to the index of action space array. I haven't run this code on my computer yet, but it seems like there might be an indexing issue doing so, (e.g. at episode = 0 (result list is empty) and best_action_index is randomly selected as 2 (selecting (75, 25) tuple), the program will not find the results at index 2 since it is empty).

also, i think that the best action index should be selected from the results of past performances (or arbitrarily for the first episode of MC) of each state action pair. e.g. if im currently at state 01, the best action should be the action that gave me the best reward at state 01. this means we might need to keep track of the state action pair rewards and use the MC update rule as we continue the training procedure. we can probably repurpose the results array-dict to be a state action pair array-dict instead.

it might be better to update the best action index within each step as well, instead of at the end of each episode

please let me know what u guys think tho, not too sure myself either

File: models/MonteCarlo.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+        nasdaq_train = train_data['NASDAQ_Returns'].values
+        msci_train = train_data['MSCI_Returns'].values
+        nasdaq_test = test_data['NASDAQ_Returns'].values
+        msci_test = test_data['MSCI_Returns'].values
+
+        return nasdaq_train, msci_train, nasdaq_test, msci_test
+
+
+    def train(self, episodes=None):
```
Comment (EdTengHR) [inline models/MonteCarlo.py]: episodes = None can be removed as it seems to be redundant

File: models/MonteCarlo.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    def __init__(self, cleaned_data, actions, n_episodes=1000, epsilon=0.1):
+        super().__init__(state_space=None, action_space=actions)
+        self.cleaned_data = cleaned_data
+        self.actions = actions
+        self.n_episodes = n_episodes
+        self.epsilon = epsilon
+        self.nasdaq_train, self.msci_train, self.nasdaq_test, self.msci_test = self.split_data()
+    
+    @staticmethod
+    def calculate_reward(action, nasdaq_return, msci_return):
```
Comment (EdTengHR) [inline models/MonteCarlo.py]: could we use the environment class' get reward function? if im remembering correctly we're not using the portfolio return as a reward, but rather the log return / sharpe ratio

File: models/MonteCarlo.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+            results.append({
+                'episode': episode,
+                'final_portfolio_value': portfolio_value,
+                'average_reward': avg_reward,
+            })
+
+        return pd.DataFrame(results), self.actions[best_action_index]
+    
+    def test(self, optimal_action):
```
Comment (EdTengHR) [inline models/MonteCarlo.py]: I think the optimal action here should be evaluated on a state-action pair by state-action pair basis right? it's not a case of finding the singular best action from training and always using that same action during testing, which is what this code seems to be doing

**Review by EdTengHR — COMMENTED** [review]
wrote a few of my thoughts, please let me know what u guys think as well. much appreciated!

File: models/MonteCarlo.py (hunk 5 above)
Comment (yuhsuanyang) [reply models/MonteCarlo.py]: I think it is better to pass episode here rather than set is as a member data in `__init__()`

File: models/MonteCarlo.py (hunk 4 above)
Comment (yuhsuanyang) [reply models/MonteCarlo.py]: I agree with @EdTengHR 's comment. A state-action pair dict or array should be kept in this class. When the env gives a state, search in the station-action pair and find the action with the best reward so far.

Besides, the paper mentioned they "Randomly generate the initial state (for each episode) beginning with the first period to period k−4 because using the same initial state may lead to memorization of the training dataset
and poor generalization". This should be implemented in the model. I have done it in the `gradient_decent_TD` branch. You may refer to it.

The base model class have defined some methods to be overridden. Aren't you going to use them?

Comment (IanAndersenNg) [reply models/MonteCarlo.py]: - I agree with this comment ill fix it and test more of my model.
- I am going to over ride them definitely, just havent done that yet.

I still have to test more results on this PR. Thanks for the feedback.

File: models/Model.py (hunk 1 above)
Comment (IanAndersenNg) [reply models/Model.py]: agreed

File: models/MonteCarlo.py (hunk 3 above)
Comment (IanAndersenNg) [reply models/MonteCarlo.py]: agreed

File: models/montecarlo.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+            return random.randint(0, len(self.actions) - 1)  # Explore
+        return np.argmax(self.calculate_average_reward(state)) # Exploit
+
+    def update_policy(self, state, action, reward):
+        reward_sum, count = self.state_action_dict[(state, action)]
+        self.state_action_dict[(state, action)] = (reward_sum + reward, count + 1)
+
+    def train(self, episodes, train_env):
+        n_steps = len(train_env.data)
+
```
Comment (yuhsuanyang) [inline models/montecarlo.py:42]: To have a better generalization as stated in the paper, please use random starting points to create each episode. You can see my implementation as a reference.

File: models/montecarlo.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        for episode in range(episodes):
+            for t in range(n_steps):
+                state = train_env.get_state(index=t)
+
+                action_index = self.choose_action(state)
+                action = self.actions[action_index]
+
+                reward = train_env.get_reward(action, index=t)
```
Comment (yuhsuanyang) [inline models/montecarlo.py]: The reward should depend on the next state at described in the paper.

File: models/montecarlo.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+- Episode: A complete run of the model from the start to the end of the training data.
+- Epsilon: The probability of choosing a random action instead of the best action for exploring instead of exploiting.
+- State-Action Dictionary: A dictionary that stores the average reward and count of each state-action pair.
+
+'''
+
+class MonteCarlo(Model):
+    def __init__(self, cleaned_data, actions, epsilon=0.1):
+        super().__init__(state_space=None, action_space=actions)
```
Comment (yuhsuanyang) [inline models/montecarlo.py]: The state space should be pass to the `super().__init__()` instead of writing in the initialize policy directly.

File: models/montecarlo.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+- State-Action Dictionary: A dictionary that stores the average reward and count of each state-action pair.
+- Reward: The reward is the next day reward of the action taken on the current day.
+- States: ['11', '10', '01', '00']
+- Actions: [(0, 100), (25, 75), (50, 50), (75, 25), (100, 0)]
+'''
+
+class MonteCarlo(Model):
+    def __init__(self, cleaned_data, actions, epsilon=0.0, state_space = ['11', '10', '01', '00']):
+        super().__init__(state_space = state_space, action_space = actions)
+        self.cleaned_data = cleaned_data
```
Comment (yuhsuanyang) [inline models/montecarlo.py]: `cleaned_data` seems to be unused.

File: models/montecarlo.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+        self.state_action_dict = self.initialize_policy()
+
+    def initialize_policy(self):
+        return {(state, action): (0,0) 
+                for state in self.state_space
+                for action in range(len(self.action_space))}
+
+    def choose_action(self, state):
+        if np.random.rand() < self.epsilon:
+            return random.randint(0, len(self.action_space) - 1)  # Explore
```
Comment (yuhsuanyang) [inline models/montecarlo.py:32]: I think it is better to move the epsilon greedy to method `train` so that this function can be called in method `test`. You can see other model implementation on the `test_result` branch.

File: models/montecarlo.py [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+                action_index = self.choose_action(state)
+                action = self.action_space[action_index]
+
+                reward = train_env.get_reward(action, index =  t + 1)
+                self.update_policy(state, action_index, reward)
+        
+        print("State action dict after all episodes: ")
+        print(self.pretty_print_state_action_dict(self.state_action_dict))
+
+        optimum_action_dict = {
```
Comment (yuhsuanyang) [inline models/montecarlo.py]: This can be a member data `self.optimum_action_dict` so that you don't need to pass it to the method `test`

File: models/montecarlo.py (hunk 12 above)
Comment (IanAndersenNg) [reply models/montecarlo.py:32]: I think we dont need to call this at the test method because we ady have the optimum map of state -> action. 

we only choose actions when we train so i think its fine as it is ?

**Comment by yuhsuanyang** [comment]
Nothing needs change.

**Review by yuhsuanyang — APPROVED** [review]

File: main.py [hunk 14]
```diff
@@ -0,0 +1,21 @@
+from environment import Environment, preprocess_data
```
Comment (EdTengHR) [inline main.py:1]: should we keep our model execution within each model's class, or have a single main class to run every model?

**Review by EdTengHR — COMMENTED** [review]
overall LGTM and I think can be merged as is, but i think we should also try to standardize the model execution flow. other models run the model directly within their own files, but MC runs from main.py. alternatively, we could also have main as a driver file to run every model

File: main.py (hunk 14 above)
Comment (yuhsuanyang) [reply main.py:1]: I think both is fine if you want to directly run the model file.
